### ISSUE TITLE
Ability to only apply action to attachments matching a regex

### ISSUE THREAD

**Issue by eyalroz** [RealRaven2000/FiltaQuilla#50]
The Thunderbird filter system applies to messages, not attachments. However, given a message on whose attachments we wish to act (e.g. delete, save or detach) - it is often the case we only care about some of these attachments - or none at all.

Specifically, in my case, I want to save all of the PDFs attached to a message, but no other attachment. I currently cannot do that (AFAICT).

This is _not_ the same issue as #5.

**Comment by RealRaven2000** [comment]
The main problem here is that attachment info is not stored anywhere in the header, IIRC there is no metadata and one has to at least stream / parse the start of the message (which may fail if it is big). What irks me about this that it is also necessary for finding out attachment name / file type etc.

Having said that I believe it may be easier to do actions async (following a synchronous system) than making conditions that work within a synchronous framework. So there may be a workaround here.

And you are right #5 is about a condition. This one is about an action, which is fundamentally different. I just can't remember whether the condition "has attachments" is pulled from the header (but you can put it into a condition anyway) then it shouldn't be impossible to do an action with the attachments asynchronously.

**Comment by eyalroz** [comment]
My logic says that if you can save it, you can, well, choose not to save it...

**Comment by RealRaven2000** [comment]
> 
> 
> My logic says that if you can save it, you can, well, choose not to save it...

exactly! And you can stick into a setTimout or async ...

**Comment by a-s-russo** [comment]
I'd also appreciate similar functionality to save attachments with a filename matching a particular pattern only, as opposed to only those files of a particular type as per the use case for @eyalroz.

For example, all attachments starting with 'Statement -' (for files named 'Statement - 345', 'Statement - 346', etc. attached to the email) to exclude other files not similarly named from being saved (for files named say 'Newsletter' etc. also attached to the email).

**Comment by RealRaven2000** [comment]
I am still in the process of fixing detaching right now (after Thunderbird removed a function that was added by Kent James on behalft of FiltaQuilla years ago), once that is fixed we can look into adding more functionality to attachments (such as renaming, maybe creating folders etc.). The main problem isn't so much implementation of more functions but a way to store the additional parameters in the filter action, any filter action has a single parameter only (in save attachments it's the stored file path of the folder where to store the attachments). And you cannot build more complex "multi-actions" because actions do not "communicate" with each other.

this means any additional functions have to be piggy-backed into the string, similar to a command line. Which also makes it less accessible to the ordinary user unless they know the command line parameters. in your case it could be something like this:

```log
  <path> -include:/regex/
or
  <path> -exclude:/regex/
```

I also found out that we cannot detach attachments from signed messages because it would break them, so I need to add an option to "save attachment" instead. I just added #373 for this.

Some work on attachments that needs to be prioritized once detachment works:

- #372
- #369
- #251 
- #219
- #43

Some of these will also require additional parameters to be baked into the "path" variable, so we will have to come up with a command line structure here, and possibly a command building interface later.

**Comment by a-s-russo** [comment]
Thanks for your prompt and informative reply. Good to know. Great add-on by the way.

Can I also please clarify the following? I've started experimenting with the 'Save attachment' functionality, and noticed that it seems to replace some characters and shorten long filenames. Am I correct in this assertion?

For example, I had an attachment with a filename that included a '#' character, and another that was more than about 60 characters in length. In Thunderbird, if I manually save the attachments (by right-clicking on them in the email and selecting 'Save As...'), the character and filename lengths are preserved. But when I use FiltaQuilla's feature, it seems to remove and shorten as explained.

(I noticed on another attachment with '/' that Thunderbird/Windows would replace it with '_' automatically when I manually saved as, but it allows '#', so FiltaQuilla seems more restrictive.)

**Comment by RealRaven2000** [comment]
> (I noticed on another attachment with '/' that Thunderbird/Windows would replace it with '_' automatically when I manually saved as, but it allows '#', so FiltaQuilla seems more restrictive.)

yes there are some blacklisted characters in order to stay compatible with all file systems, you can try explicitely add them back in FiltaQuilla Preferences / Actions Tab:

<img width="733" height="192" alt="Image" src="https://github.com/user-attachments/assets/557fbb7d-7797-4e9d-be16-6aab885b76a2" />

I think there are some characters that are "hardlocked" such as `/` `,` or `\` - but hopefully `#` may be able to let back in... I should probably add a tooltip that explains this feature to some level of detail.

**Comment by a-s-russo** [comment]
Thanks, that worked! Great to know I can whitelist '#'.
